mbedtls: provide user mode access

tyrantlizard77 opened the pull request:
The mbedtls library has some globals which results in faults
when user mode tries to access them.

Instantiate a memory partition for mbedtls's globals.
Add the mbedtls heap, and the function pointers used by
mbedtls's platform abstraction.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
mbedtls: provide user mode access

The mbedtls library has some globals which results in faults
when user mode tries to access them.

Instantiate a memory partition for mbedtls's globals.
The linker will place all globals found by building this
library into this partition.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
Note to reviewers: this patch required modifying the upstream mbedtls code, in order to correctly place the globals for the platform abstraction. This is necessary since the mbedtls library is built from source by the kernel.

An alternative approach could be to build mbedtls as a .a library archive, and then add a directive like `-l libmbedtls.a k_mbedtls_partition` to the invocation of gen_app_partitions.py. This will have the same effect without needed any K_APP_DMEM() tags in the mbedtls code. I'm not sure if this is possible in our build system currently.

andrewboie commented:
Also note: x86 will be failing CI until #13918 is fixed.

agross-linaro commented:
If I had my druthers, I'd prefer the .a path.  Any changes we make to external libraries makes it more costly to us to maintain.

andrewboie commented:
@agross-linaro agreed, and I figured it out, see updated patch

ceolin reviewed (commented):
current solution is better IMHO, but the other could be change, those globals are used when the platform does not define those functions, we could add them in  custom MBEDTLS_CONFIG_FILE (as we already do for other things). The problem is that we would need to do it for all available configurations.

This solution is just cleaner to me.

andrewboie commented:
> current solution is better IMHO, but the other could be change, those globals are used when the platform does not define those functions, we could add them in custom MBEDTLS_CONFIG_FILE (as we already do for other things). The problem is that we would need to do it for all available configurations.

I'm not sure I understand your review.
Let's just focus on the current implementation, I see no downsides to it.

andrewboie commented:
I'm seeing an unexpected failure in tests/subsys/jwt:

```
***** Booting Zephyr OS v1.14.0-rc1-624-ge68e0ac0f8d1 *****
Running test suite lib_jwt_test
===================================================================
starting test - test_jwt

    Assertion failed at /home/apboie/projects/zephyr3/tests/subsys/jwt/src/main.c:52: test_jwt: (res not equal to 0)
Signing payload
FAIL - test_jwt
===================================================================
Test suite lib_jwt_test failed.
```

Trying to discover what is going on.

Commit:
userspace: zero app memory bss earlier

Some init tasks may use some bss app memory areas and
expect them to be zeroed out. Do this much earlier
in the boot process, before any of the init tasks
run.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andyross commented on include/app_memory/partitions.h:
```diff
@@ -0,0 +1,17 @@
+#include <kernel.h> /* For struct k_mem_partition */
+
+#if defined(CONFIG_MBEDTLS)
+extern struct k_mem_partition k_mbedtls_partition;
```
Is mbedtls enough of a core thing that it belongs in a single global header like this?  Seems aesthetically that it really ought to be exported by a mbedtls header.  Meh.

andrewboie replied:
mbedtls headers are from upstream project. we don't want to change those, this is very specific to Zephyr.

I'm going to put other partitions in here too as we create them.

ceolin replied:
There are headers that we can change to tuning mbedTLS options in  ./ext/lib/crypto/mbedtls/configs.
Though I've no complain with this new header since it will be used by other partitions.

andrewboie replied:
> There are headers that we can change to tuning mbedTLS options in ./ext/lib/crypto/mbedtls/configs

It wouldn't be appropriate to define this there.
For one thing, there are a lot of different config headers and we'd have to change all of them. People creating new config headers could easily trip over this.
.

agross-linaro replied:
@ceolin the main config files are generally ones you don't want to mess with.  You can fine tune those by adding your own application specific overrides.

ceolin replied:
well, some of these configs contains changes to allows fine tuning on build. Some of them like `config-tls-generic.h` does not even exist in mbedTLS source. What I'm missing ?

Anyway, I don't want bring noise to this PR because this implementation is better IMHO :)

agross-linaro replied:
@ceolin that was just an off the cuff statement.  This reminded me of the mqtt demo sample we added where I had specified specific config overrides in the application source directory.

andyross reviewed (approved):
Nitpick about scoping of symbols.  Looks sane to me.

andrewboie commented:
I can't seem to reproduce the two other CI failures:

jukkar reviewed (approved):
LGTM

andrewboie commented:
recheck

andrewboie commented on ext/lib/crypto/mbedtls/zephyr_init.c:
```diff
@@ -20,6 +21,8 @@
 #endif /* CONFIG_MBEDTLS_CFG_FILE */
 #endif
 
+K_APPMEM_PARTITION_DEFINE(k_mbedtls_partition);
```
bug: this needs to live in kernel memory, but since we are pulling in the .a file, this partition definition is instead living in app shared memory. need to fix this.

marc-hb commented:
Re-adding link to non-last and "unstable" Shippable run which failed with `zephyr/priv_stacks_prebuilt.elf section bss will not fit in region SRAM` 2/5 times:
https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/36138

(At first I thought it was unintended but no: hiding and/or deleting past information seems to be one of github's design choices. Not nice.)
